### PR TITLE
Assignment7+8 linq

### PR DESCRIPTION
**Pull request by DaisyPalencia**
Paired with @AnnikaStachowiak.

### CODE REVIEW

**Comment by Malcolm-Shepherd** [comment]
I'm going to review this

File: Assignment/Assignment.Tests/SampleDataTests.cs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+            SampleData sampleData = _data;
+
+            var people = sampleData.People;
+
+            Assert.AreEqual(50, people.Count());
+
+
+        }
+        [TestMethod]
+		public void SampleData_GetUniqueSortedListOfStates_Success()
```
Comment (Malcolm-Shepherd) [inline Assignment/Assignment.Tests/SampleDataTests.cs:37]: This doesn't check if the whole list is sorted and doesn't use linq

File: Assignment/Assignment.Tests/SampleDataTests.cs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            SampleData sampleData = _data;
+
+            IEnumerable<string> state = sampleData.GetUniqueSortedListOfStatesGivenCsvRows();
+
+			Assert.AreEqual<string>("AL", state.First());
+			Assert.AreEqual<string>("WV", state.Last());
+
+        }
+        [TestMethod]
+        public void SampleData_HardCodedAddresses_Success()
```
Comment (Malcolm-Shepherd) [inline Assignment/Assignment.Tests/SampleDataTests.cs:48]: This doesn't test any methods we created.

File: Assignment/Assignment/SampleData.cs [hunk 3]
```diff
@@ -1,30 +1,75 @@
...
-        public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() 
-            => throw new NotImplementedException();
-
+        public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()
+        {
+            return CsvRows.Select(row => row.Split(',')[6]).Distinct().OrderBy(state => state).ToList();
+
+        }
         // 3.
         public string GetAggregateSortedListOfStatesUsingCsvRows()
```
Comment (Malcolm-Shepherd) [inline Assignment/Assignment/SampleData.cs:37]: No test to show this works as expected.

File: Assignment/Assignment/SampleData.cs [hunk 4]
```diff
@@ -1,30 +1,75 @@
...
-
-        // 5.
-        public IEnumerable<(string FirstName, string LastName)> FilterByEmailAddress(
-            Predicate<string> filter) => throw new NotImplementedException();
-
+        public IEnumerable<IPerson> People
+        {
+            get
+            {
+                IEnumerable<Person> people = from person in CsvRows.OrderBy(state => state.Split(',')[6])
```
Comment (Malcolm-Shepherd) [inline Assignment/Assignment/SampleData.cs:48]: 2 of the requirements for this aren't tested. I'm pretty sure it works as intended, but I'm not sure.

**Review by Malcolm-Shepherd — COMMENTED** [review]
1. Implement the `ISampleData.CsvRows` property, loading the data from the `People.csv` file and returning each line as a single string. ✔

   - Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
   - Using LINQ, skip the first row in the `People.csv`. ✔
   - Be sure to appropriately handle resource (`IDisposable`) items correctly if applicable (and it may not be depending on how you implement it). ✔

2. Implement `IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()` to return a **sorted**, **unique** list of states. ❌✔

   - Use `ISampleData.CsvRows` for your data source. ✔
   - Don't forget the list should be unique. ✔
   - Sort the list alphabetically. ✔
   - Include a test that leverages a hardcoded list of addresses. ❌
   - Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ❌

3. Implement `ISampleData.GetAggregateSortedListOfStatesUsingCsvRows()` to return a `string` that contains a **unique**, comma separated list of states. ❌✔

   - Use `ISampleData.GetUniqueSortedListOfStatesGivenCsvRows()` for your data source. ✔
   - Consider "selecting" only the states and calling `ToArray()` to retrieve an array of all the state names. ✔
   - Given the array, consider using `string.Join` to combine the list into a single string. ✔

4. Implement the `ISampleData.People` property to return all the items in `People.csv` as `Person` objects ✔

   - Use `ISampleData.CsvRows` as the source of the data. ✔
   - Sort the list by State, City, and Zip. (Sort the addresses first then select). ❌✔
   - Be sure that `Person.Address` is also populated. ❌✔
   - Adding null validation to all the `Person` and `Address` properties is **optional**.
   - Consider using `ISampleData.CsvRows` in your test to verify your results. ✔

5. Implement `ISampleDate.FilterByEmailAddress(Predicate<string> filter)` to return a list of names where the email address matches the `filter`. ✔

   - Use `ISampleData.People` for your data source. ✔

6. Implement `ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people)` to return a `string` that contains a **unique**, comma-separated list of states. ❌✔

   - Use the `people` parameter from `ISampleData.People` property for your data source. ✔
   - At a minimum, use the `System.Linq.Enumerable.`Aggregate` LINQ method to create your result. ✔
   - Don't forget the list should be unique. ✔
   - It is recommended that, at a minimum, you use `ISampleData.GetUniqueSortedListOfStatesGivenCsvRows` to validate your result.

7. Given the implementation of `Node` in Assignment5

- Implement `IEnumerable<T>` to return all the items in the "circle" of items. ✔
- Add an `IEnumberable<T> ChildItems(int maximum)` method to `Node` that returns the remaining items with a maximum number of items returned less than `maximum`.  ✔

## Fundamentals

- Place all shared project properties into a `Directory.Build.Props` file.
- Place all shared project items into a `Directory.Build.targets` file.
- Ensure nullable reference types is enabled  ✔
- Ensure that you turn on code analysis for all projects(EnableNETAnalyzers)  ✔
- Set `LangVersion` and the `TargetFramework` to the latest released versions available (preview versions optional)   ✔
- and enabled .NET analyzers for both projects ✔
- For this assignment, consider using `Assert.AreEqual<T>()` (the generic version)  ❌
- All of the above should be unit tested ❌
- Choose simplicity over complexity ✔

**Comment by JessicaH46** [comment]
I will review this

File: Assignment/Assignment.Tests/SampleDataTests.cs (hunk 1 above)
Comment (AnnikaStachowiak) [reply Assignment/Assignment.Tests/SampleDataTests.cs:47]: Noted

File: Assignment/Assignment.Tests/SampleDataTests.cs (hunk 2 above)
Comment (AnnikaStachowiak) [reply Assignment/Assignment.Tests/SampleDataTests.cs]: Changed

File: Assignment/Assignment/SampleData.cs (hunk 3 above)
Comment (AnnikaStachowiak) [reply Assignment/Assignment/SampleData.cs:37]: Noted

File: Assignment/Assignment/SampleData.cs (hunk 4 above)
Comment (AnnikaStachowiak) [reply Assignment/Assignment/SampleData.cs:48]: Thanks for the feedback

File: Assignment/Assignment/Node.cs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    {
+        var node = Next;
+        while (node.Next != this)
+        {
+            node = node.Next;
+        }
+        node.Next = Next;
+        Next = this;
+    }
+        public IEnumerable<T> ChildItems(int maximum)
```
Comment (mmwoodfo) [inline Assignment/Assignment/Node.cs:76]: You should have some tests for this method!

**Review by mmwoodfo — COMMENTED** [review]
Good job overall!

**Pull request closed without merging**